### DISCUSSION TITLE
Using Generics without stripping fields

### DISCUSSION THREAD

**Discussion by andy-gorman-dandy** [colinhacks/zod#1659] [Q&A]
I'm currently trying to infer a type as an interface using the code below.

```typescript
function CartItemUnitMetadataSchema<Type extends string = string>(unitType: Type) {
    return z.object({
        unit_type: z.literal(unitType),
        material: z.string().nullish()
    });
}

interface ICartItemUnitMetadata<Type extends string = string> extends z.infer<ReturnType<typeof CartItemUnitMetadataSchema<Type>>> {}
```

However, I get the following typescript error
`TS2312: An interface can only extend an object type or intersection of object types with statically known members.`

If I change the `CartItemUnitMetadataSchema` function to the following, I'm able to use the code without errors, but lose the generics I'm trying to get.

```typescript
function CartItemUnitMetadataSchema<Type extends string = string>(unitType: Type) {
    return z.object({
        unit_type: z.literal<string>(unitType),
        material: z.string().nullish()
    });
}
```

It seems to me that when I'm using generics in the first example, something in zod is stripping out those fields and possibly making them optional? I've seen some issues/discussions seemingly around this same question, but don't have a lot of clarity on what's happening based on what I've seen so far. If anyone could provide any insight that would be tremendously helpful. Thank you!

**Comment by JacobWeisenburger** [comment]
Everything seems to be working as expected. See below. Is there something I'm missing?

```ts
function CartItemUnitMetadataSchema<Type extends string = string> ( unitType: Type ) {
    return z.object( {
        unit_type: z.literal( unitType ),
        material: z.string().nullish()
    } )
}

const schema = CartItemUnitMetadataSchema( 'foo' )
const data = schema.parse( { unit_type: 'foo' } )
type Data = z.infer<typeof schema>
// type Data = {
//     material?: string | null | undefined
//     unit_type: 'foo'
// }

console.log( data ) // { unit_type: 'foo' }
```

**Reply by andy-gorman-dandy** [reply]
I should have been a little clearer in my original post. We have an existing interface that's quite common in our codebase, defined as 
```typescript
interface ICartItemUnitMetadata<Type string extends string = string> {
  unit_type: Type;
  material?: string;
}
```

And using the schema function above, I'd like to be able to redefine that interface as 
```ts
interface ICartItemUnitMetadata<Type extends string = string> extends z.infer<ReturnType<typeof CartItemUnitMetadataSchema<Type>>> {}
```
and not have to change any of the places we implement that interface.

This does work if I define `unit_type: z.literal<string>(unitType)`, but then I lose the generic typing I'm trying to get.

**Comment by JacobWeisenburger** [comment]
I changed the `interface` to a `type` and it seems to work for me.

Let me know if I'm missing something.

```ts
type ICartItemUnitMetadata<Type extends string = string> =
    z.infer<ReturnType<typeof CartItemUnitMetadataSchema<Type>>>
```

Full Example:
```ts
function CartItemUnitMetadataSchema<Type extends string = string> ( unitType: Type ) {
    return z.object( {
        unit_type: z.literal( unitType ),
        material: z.string().nullish()
    } )
}

type ICartItemUnitMetadata<Type extends string = string> =
    z.infer<ReturnType<typeof CartItemUnitMetadataSchema<Type>>>

const schema = CartItemUnitMetadataSchema( 'foo' )
const data = schema.parse( { unit_type: 'foo' } )
type Data = z.infer<typeof schema>
// type Data = {
//     material?: string | null | undefined
//     unit_type: 'foo'
// }

console.log( data ) // { unit_type: 'foo' }
```

**Reply by andy-gorman-dandy** [reply]
@JacobWeisenburger I have done the same, but a lot of our code needs to `implement ICartItemUnitMetadata`, which it can't do when it's typed the way you showed here. I think I found the problem though and am considering opening an issue for it, though I don't know the fix offhand.

Zod object typing uses the following types
```typescript
    type optionalKeys<T extends object> = {
        [k in keyof T]: undefined extends T[k] ? k : never;
    }[keyof T];
    type requiredKeys<T extends object> = {
        [k in keyof T]: undefined extends T[k] ? never : k;
    }[keyof T];
```

If `T[k]` is a generic type, `undefined extends T[k]` does not evaluate to true or to false, so those keys are omitted from this type altogether. This seems like a typescript level limitation, so I'm not certain to as to what the workaround would be, but it would explain the behavior that I am seeing.

**Reply by andy-gorman-dandy** [reply]
After mulling and reading about [this issue](https://github.com/microsoft/TypeScript/issues/31751), I think this can be resolved by updating the above code to the following.

I will play around with this and see if I can open a PR with those changes
```typescript
    type optionalKeys<T extends object> = {
        [k in keyof T]: [undefined] extends [T[k]] ? k : never;
    }[keyof T];
    type requiredKeys<T extends object> = {
        [k in keyof T]: [undefined] extends [T[k]] ? never : k;
    }[keyof T];
```

**Reply by JacobWeisenburger** [reply]
You might not be able to in your project, but I try to use types instead of interfaces as much as possible. I find types to be more reliable and not have as many problems as interfaces. However, this might just work well for me because I mostly do Functional Programming. I find OOP to have many problems that I don't have to worry about when using FP.
